### PR TITLE
gpt-auto-generator: enable TPM2 unlocking in gpt-auto-generator

### PR DESCRIPTION
**Pull request by poettering**
If we detect a TPM, let's also unlock the disk with it, if it has an enrollment for that.

See: #30176

### CODE REVIEW

**Comment by poettering** [comment]
So this should fix the specific issue in #30176, and we probably should merge this, before v255.

But we actually have a bigger problem in general, and I am not sure how to address this best: since token module support was added to cryptsetup we'd try a "blanket" activation via token before doing the finer grained stuff. Butt that's a mess, because it will pull in our own token modules, and run them before we get to do so explicitly. This is a mess because this way we don't get the chance to pass our various settings into the token modules, hence they run without...

We probably should tweak the blanket activation to do so only for tokens that are not ours, i.e. foreign.

Man this is all so complex...

**Comment by thesamesam** [comment]
That fixes it for me, thanks.

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 1]
```diff
@@ -100,21 +100,27 @@ static int add_cryptsetup(
...
                         if (!strextend_with_separator(&options, ",", "tpm2-measure-pcr=yes"))
                                 return log_oom();
                 }
         }
 
+        if (uki > 0) {
+                /* Enable TPM2 based unlocking automatically, if we have a TPM. */
+                if (!strextend_with_separator(&options, ",", "tpm2-device=auto"))
+                        return log_oom();
+        }
```
Comment (YHNdnzj) [inline src/gpt-auto-generator/gpt-auto-generator.c]: It is easier to follow if this is placed before the if branch for `measure`. Also, brackets are unnecessary.

**Review by YHNdnzj — APPROVED** [review]
Just some cosmetic stuff

**Comment by SoapGentoo** [comment]
any reason not to merge this?

**Review by bluca — APPROVED** [review]

**Comment by gdamjan** [comment]
Ah nice,
this should've been mentioned in the NEWS file.

Goodbye setting `rd.luks.name=b598…=root rd.luks.options=b598…=tpm2-device=auto` in my kernel command line. this was the only place where I had to specify the boot device in the UKI world.

**Comment by ElvishJerricco** [comment]
If I'm understanding correctly, the way `efi_measured_uki` is used here means this only works if you boot with a sd-stub UKI. I mean I guess it can be argued that trying to TPM-unlock the root fs from a Type 1 boot entry is maybe not the best idea, but it still seems like it would make sense to have. At the very least, a TPM+pin LUKS slot is more secure than just the pin.

**Comment by VannTen** [comment]
> if it has an enrollment for that.

From looking at the diff and the code, it does not appear to check whether the LUKS header contain a tpm token at all, only if a TPM exists ?

AFAICT, this regressed https://github.com/systemd/systemd/issues/32586 (aka unlocking the root with a FIDO2 token). (adding some comments on the issue as well)

**Pull request merged**